### PR TITLE
Add rule for testing if key is set to an empty (null) value

### PR DESCRIPTION
**Pull request by gdubicki**
Implements #82

### CODE REVIEW

**Comment by gdubicki** [comment]
This is my first PR here, so feedback is most welcome. :)

**Comment by gdubicki** [comment]
Thanks @adrienverge !

After I slept over with this I woke up with similar thoughts about more tests being needed and what would and what should happen with empty sequences and maps.

I will try to improve my code in the next few days, but if you (or anyone else!) want to do it, then of course feel free. Just let me know here when you start so we won't duplicate the effort.

File: yamllint/conf/default.yaml [hunk 1]
```diff
@@ -32,6 +32,7 @@ rules:
     max: 2
     max-start: 0
     max-end: 0
+  empty-values: enable
```
[the remark this answers is not in the document]
Comment (gdubicki) [reply yamllint/conf/default.yaml]: Ok, I will change it.

**Comment by adrienverge** [comment]
Great!

I also thought a bit more about it, and I'm not sure what the best implementation would be.

Do we want this rule to avoid *visually empty* values (case A)? or to forbid `null` values (case B)?

To be clearer, for case A I would expect:

```yaml
null1:               # NOT OK
null2: null          # OK, it's null but it's explicit
null3: ''            # OK, it's empty but it's explicit
map: {a: , b:, c: }  # NOT OK
seq:
  -                  # NOT OK
  - null             # OK, it's null but it's explicit
  - []               # OK, it's empty but it's explicit
```

... and for case B I would expect:

```yaml
null1:               # NOT OK
null2: null          # NOT OK
null3: ''            # NOT OK if allow-empty-string: false
map: {a: , b:, c: }  # NOT OK
seq:
  -                  # NOT OK
  - null             # NOT OK
  - []               # NOT OK if allow-empty-sequence: false
```

I have the feeling that case A would be more useful than B: it would avoid confusios like described in #82. And I'm not sure case B (strictly forbidding `null` values) would help anyone. Also for case A, my 'allow-empty-string', 'allow-empty-sequence', etc. propositions don't make much sense.

What do you think?

**Comment by freddrake** [comment]
I would find case A useful, but not case B.  Implied-null is something I really dislike about YAML, and don't want it in my projects.

**Comment by gdubicki** [comment]
That's right, for my use case I only want to detect empty values that could be caused by incorrect indentation, so:

```
null1:         # NOT OK, because this null value comes from line ending with ":" 
#                and possibly because line below like "foo: bar" is not being indented correctly
null2: null    # OK, it's null but it's explicit
null3: ''      # OK, it's empty but it's explicit

map1: {}       # OK, it's an empty map but it's explicit
map1: {a: }    # OK, it's null but it's explicit because line does not end with ":"
map2:
  a:           # NOT OK, because this null value comes from line ending with ":"
#                and possibly because line below like "foo: bar" is not being indented correctly
```

But frankly speaking I don't know what to do with an empty seq value:

```
seq:
  -           # <--- OK or not ???
  - null      # OK, it's null but it's explicit
  - []        # OK, as this empty seq is explicit
```

...because it feels like this can be a problem, but I can't come up with an obvious and simple way this mistake could be made.

What do you think?

**Comment by adrienverge** [comment]
Agreed, so we want to avoid *non-explicit empty values*, because they implicitly fallback to `null`.

I understand your use-case is only about:
```yaml
mapping:
sub-element: incorrectly indented
```
... but I suspect other ones (like `map: {a: , b:, c: }` or lists with empty items) can come in the future.

So I suggest this configuration for the rule:
```yaml
empty-values:
  forbid-in-block-mappings: true
```
This way, in the future we could extend it to:
```yaml
empty-values:
  forbid-in-block-mappings: true   # your use-case
  forbid-in-flow-mappings: true    # mapping: {a: , b:, c: }
  forbid-in-block-sequences: true  # sequence:\n  -\n  - item
  forbid-in-flow-sequences: true   # sequence: [ , , ]
```

Does this sounds good to you?

If yes, your PR wouldn't need so much change, except testing many more cases to make sure this new rule behaves as intended in corner cases.

**Comment by gdubicki** [comment]
Sounds good. I will try to implement this in the next few days.

**Comment by gdubicki** [comment]
Implementation updated for new config format, tests added. Does it look good to you now, @adrienverge ?

PS `sequence: [ , , ]` was reported as syntactically incorrect, so I am not sure if we are going to have `forbid-in-flow-sequences`...

**Comment by adrienverge** [comment]
PS: I have started to implement `forbid-in-flow-mappings`, I'll add it once your PR is merged ;-)
https://github.com/adrienverge/yamllint/commits/wip/empty-values-forbid-in-flow-mappings

File: docs/rules.rst [hunk 2]
```diff
@@ -59,6 +59,11 @@ empty-lines
 
 .. automodule:: yamllint.rules.empty_lines
 
+empty-values
+-----------
```
[the remark this answers is not in the document]
Comment (gdubicki) [reply docs/rules.rst]: Fixed.

File: tests/rules/test_indentation.py [hunk 3]
```diff
@@ -1316,7 +1316,8 @@ def test_multi_lines(self):
                    '...\n', conf)
 
     def test_empty_value(self):
-        conf = 'indentation: {spaces: consistent}'
+        conf = ('indentation: {spaces: consistent}\n'
+                'empty-values: disable\n')
```
[the remark this answers is not in the document]
Comment (gdubicki) [reply tests/rules/test_indentation.py]: Fixed.

File: yamllint/rules/empty_values.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+* Use ``forbid-in-block-mappings`` to prevent implicit empty values in block
+  mappings.
+
+.. rubric:: Examples
+
+#. With ``empty-values: {forbid-in-block-mappings: true}``
+
+   the following code snippet would **PASS**:
+   ::
```
[the remark this answers is not in the document]
Comment (gdubicki) [reply yamllint/rules/empty_values.py]: Fixed. (Btw: it would be good to add building docs step to Travis so that contributors like me would have both code and docs tested after push.)

File: tests/rules/test_indentation.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (gdubicki) [reply tests/rules/test_indentation.py]: Fixed.

File: yamllint/conf/relaxed.yaml [hunk 5]
```diff
@@ -18,6 +18,8 @@ rules:
   document-start: disable
   empty-lines:
     level: warning
+  empty-values:
+    forbid-in-block-mappings: false
```
[the remark this answers is not in the document]
Comment (gdubicki) [reply yamllint/conf/relaxed.yaml]: Fixed.

File: docs/rules.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (gdubicki) [reply docs/rules.rst]: Fixed.

**Comment by gdubicki** [comment]
All requested fixes done, please review @adrienverge . :)

**Comment by gdubicki** [comment]
It would be nice if you could release a new version with this rule quickly, as we are eager to start using it. :)

File: yamllint/rules/empty_values.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (adrienverge) [reply yamllint/rules/empty_values.py]: You're totally right. I will do!

**Comment by adrienverge** [comment]
Almost perfect!

The docs compilation still failed on a warning, and examples were not represented as code (`::` blocks must be followed by an empty line to work).

So I allowed myself to make some changes and `push --force` to your branch. Hope that's OK.

```diff
--- a/yamllint/rules/empty_values.py
+++ b/yamllint/rules/empty_values.py
@@ -15,33 +15,35 @@
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
 """
-Use this rule to prevent nodes with empty content, that implicitly result
-in null values.
+Use this rule to prevent nodes with empty content, that implicitly result in
+``null`` values.
 
 .. rubric:: Options
 
 * Use ``forbid-in-block-mappings`` to prevent implicit empty values in block
-mappings.
+  mappings.
 
 .. rubric:: Examples
 
 #. With ``empty-values: {forbid-in-block-mappings: true}``
-the following code snippets would **PASS**:
-
+   the following code snippets would **PASS**:
    ::
+
     some-mapping:
       sub-element: correctly indented
 
    ::
+
     explicitly-null: null
 
    the following code snippets would **FAIL**:
-
    ::
+
     some-mapping:
     sub-element: incorrectly indented
 
    ::
+
     implicitly-null:
 
 """
@@ -59,12 +61,9 @@ CONF = {'forbid-in-block-mappings': bool}
 def check(conf, token, prev, next, nextnext, context):
 
     if conf['forbid-in-block-mappings']:
-        if isinstance(token, yaml.ValueToken) and \
-                (isinstance(next, yaml.KeyToken)
-                 or isinstance(next, yaml.BlockEndToken)
-                 or isinstance(next, yaml.StreamEndToken)
-                 or (isinstance(next, yaml.DocumentEndToken))):
-
+        if isinstance(token, yaml.ValueToken) and isinstance(next, (
+                yaml.KeyToken, yaml.BlockEndToken,
+                yaml.StreamEndToken, yaml.DocumentEndToken)):
             yield LintProblem(token.start_mark.line + 1,
                               token.end_mark.column + 1,
-                              'empty key value in block mapping')
+                              'empty value in block mapping')
--- a/tests/rules/test_empty_values.py
+++ b/tests/rules/test_empty_values.py
@@ -101,8 +101,7 @@ class EmptyValuesTestCase(RuleTestCase):
                    'explicitly-null: null\n', conf)
 
         self.check('---\n'
-                   'explicitly-null:with-colons:'
-                   'in-key: null\n', conf)
+                   'explicitly-null:with-colons:in-key: null\n', conf)
 
         self.check('---\n'
                    'false-null: nulL\n', conf)
```

> It would be nice if you could release a new version with this rule quickly, as we are eager to start using it. :)

Sure, I'll try this week-end. I need to add `forbid-in-flow-mappings`.

**Comment by gdubicki** [comment]
Great, thanks for fixing that warning for me! My version of the doc builder didn’t complain about that..

**Comment by adrienverge** [comment]
@gdubicki It is now available in version 1.10.0!
-> https://github.com/adrienverge/yamllint/commit/a92743c8ca44bab56389c75504fcf25d33e2c11a
